### PR TITLE
WebGPU should be exposed in DedicatedWorkers

### PR DESCRIPTION
**Pull request by mwyrzykowski**
#### ff507241e7dd8c7bc6d01f11027c09e8e3f0dc1b
<pre>
WebGPU should be exposed in DedicatedWorkers
<a href="https://bugs.webkit.org/show_bug.cgi?id=232542">https://bugs.webkit.org/show_bug.cgi?id=232542</a>&gt;
&lt;radar://85108397&gt;

Reviewed by Tadeu Zagallo.

Allow WebGPU to render in a web worker.

This patch is mostly mechanical and doesn&apos;t implement copying
the buffer from OffscreenCavnas, so nothing is visible, but
GPU frame capture indicates the expected results are rendered.

* Source/JavaScriptCore/inspector/protocol/Canvas.json:
* Source/WebCore/Modules/WebGPU/GPU.idl:
* Source/WebCore/Modules/WebGPU/GPUAdapter.idl:
* Source/WebCore/Modules/WebGPU/GPUAdapterInfo.idl:
* Source/WebCore/Modules/WebGPU/GPUBindGroup.idl:
* Source/WebCore/Modules/WebGPU/GPUBindGroupLayout.idl:
* Source/WebCore/Modules/WebGPU/GPUBuffer.idl:
* Source/WebCore/Modules/WebGPU/GPUBufferUsage.idl:
* Source/WebCore/Modules/WebGPU/GPUColorWrite.idl:
* Source/WebCore/Modules/WebGPU/GPUCommandBuffer.idl:
* Source/WebCore/Modules/WebGPU/GPUCommandEncoder.idl:
* Source/WebCore/Modules/WebGPU/GPUCompilationInfo.idl:
* Source/WebCore/Modules/WebGPU/GPUCompilationMessage.idl:
* Source/WebCore/Modules/WebGPU/GPUComputePassEncoder.idl:
* Source/WebCore/Modules/WebGPU/GPUComputePipeline.idl:
* Source/WebCore/Modules/WebGPU/GPUDevice.idl:
* Source/WebCore/Modules/WebGPU/GPUDeviceLostInfo.idl:
* Source/WebCore/Modules/WebGPU/GPUExternalTexture.idl:
* Source/WebCore/Modules/WebGPU/GPUInternalError.idl:
* Source/WebCore/Modules/WebGPU/GPUMapMode.idl:
* Source/WebCore/Modules/WebGPU/GPUOutOfMemoryError.idl:
* Source/WebCore/Modules/WebGPU/GPUPipelineLayout.idl:
* Source/WebCore/Modules/WebGPU/GPUQuerySet.idl:
* Source/WebCore/Modules/WebGPU/GPUQueue.idl:
* Source/WebCore/Modules/WebGPU/GPURenderBundle.idl:
* Source/WebCore/Modules/WebGPU/GPURenderBundleEncoder.idl:
* Source/WebCore/Modules/WebGPU/GPURenderPassEncoder.idl:
* Source/WebCore/Modules/WebGPU/GPURenderPipeline.idl:
* Source/WebCore/Modules/WebGPU/GPUSampler.idl:
* Source/WebCore/Modules/WebGPU/GPUShaderModule.idl:
* Source/WebCore/Modules/WebGPU/GPUShaderStage.idl:
* Source/WebCore/Modules/WebGPU/GPUSupportedFeatures.idl:
* Source/WebCore/Modules/WebGPU/GPUSupportedLimits.idl:
* Source/WebCore/Modules/WebGPU/GPUTexture.idl:
* Source/WebCore/Modules/WebGPU/GPUTextureUsage.idl:
* Source/WebCore/Modules/WebGPU/GPUTextureView.idl:
* Source/WebCore/Modules/WebGPU/GPUUncapturedErrorEvent.idl:
* Source/WebCore/Modules/WebGPU/GPUValidationError.idl:
* Source/WebCore/html/OffscreenCanvas.cpp:
(WebCore::OffscreenCanvas::createContextWebGPU):
(WebCore::OffscreenCanvas::getContext):
* Source/WebCore/html/OffscreenCanvas.h:
* Source/WebCore/html/OffscreenCanvas.idl:
* Source/WebCore/html/canvas/GPUCanvasContext.idl:
* Source/WebCore/html/canvas/GPUCanvasContextCocoa.h:
* Source/WebCore/html/canvas/GPUCanvasContextCocoa.mm:
(WebCore::GPUCanvasContextCocoa::htmlOrOffscreenCanvas const):
(WebCore::GPUCanvasContextCocoa::GPUCanvasContextCocoa):
(WebCore::GPUCanvasContextCocoa::prepareForDisplayWithPaint):
(WebCore::GPUCanvasContextCocoa::paintRenderingResultsToCanvas):
(WebCore::GPUCanvasContextCocoa::canvas):
(WebCore::GPUCanvasContextCocoa::markContextChangedAndNotifyCanvasObservers):
* Source/WebCore/inspector/InspectorCanvas.cpp:
(WebCore::InspectorCanvas::buildObjectForCanvas):
* Source/WebCore/page/Chrome.h:
* Source/WebCore/page/WorkerNavigator.cpp:
(WebCore::WorkerNavigator::gpu):

* Source/WebCore/page/WorkerNavigator.h:
* Source/WebCore/platform/GraphicsClient.h:
* Source/WebKit/Shared/WebGPU/WebGPUIdentifier.h:
Changed to AtomicObjectIdentifier as workers will render off the main thread.

* Source/WebKit/WebProcess/WebCoreSupport/WebWorkerClient.cpp:
(WebKit::WebWorkerClient::WebWorkerClient):
(WebKit::WebWorkerClient::clone):
(WebKit::WebWorkerClient::createGraphicsContextGL const):
(WebKit::WebWorkerClient::createGPUForWebGPU const):
* Source/WebKit/WebProcess/WebCoreSupport/WebWorkerClient.h:

Canonical link: <a href="https://commits.webkit.org/268693@main">https://commits.webkit.org/268693@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/2f0fbe1cf22797628d350884755d1732c51b1f7b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20279 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20712 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21352 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22180 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18927 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20511 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23969 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/20882 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/20362 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20499 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20403 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17636 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23030 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17572 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18439 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/24693 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/17649 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18648 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18615 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22666 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/19659 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19196 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16302 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/23686 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18401 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/5695 "Found 4 jsc stress test failures: stress/sampling-profiler-microtasks.js.bytecode-cache, stress/sampling-profiler-microtasks.js.dfg-eager, stress/sampling-profiler-microtasks.js.no-llint, wasm.yaml/wasm/stress/simple-inline-stacktrace-with-catch.js.wasm-eager") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4899 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22742 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/24943 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19021 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5505 "Passed tests") |

### CODE REVIEW

**Review by tadeuzagallo — APPROVED** [review]

File: Source/WebCore/inspector/InspectorCanvas.cpp [hunk 1]
```diff
@@ -892,6 +893,9 @@ Ref<Protocol::Canvas::Canvas> InspectorCanvas::buildObjectForCanvas(bool capture
         if (is<WebGL2RenderingContext>(m_context))
             return Protocol::Canvas::ContextType::WebGL2;
 #endif
+        if (is<GPUCanvasContext>(m_context))
+            return Protocol::Canvas::ContextType::WebGPU;
```
Comment (dcrousso) [inline Source/WebCore/inspector/InspectorCanvas.cpp:897]: why was this included as part of this PR?  i ask because this is not how WebGPU inspection was done in the past (see [215342@main](https://commits.webkit.org/215342@main)), so it's really odd for it to be done this way now

Comment (mwyrzykowski) [reply Source/WebCore/inspector/InspectorCanvas.cpp:897]: It was included because opening web inspector resulted in a crash (debug assertion) otherwise when WebGPU was being used in a web worker

Comment (dcrousso) [reply Source/WebCore/inspector/InspectorCanvas.cpp:897]: i see.  my guess is that this is because inside `InspectorCanvasAgent::internalEnable` we're iterating over all `CanvasRenderingContext::instances()` and attempting to create an `InspectorCanvas` for each of them rather than only the ones that Web Inspector currently supports (see callers of `InspectorInstrumentation::didCreateCanvasRenderingContext`)

id suggest undoing this part of the PR and instead add either
```diff
-#if ENABLE(OFFSCREEN_CANVAS)
-            if (is<PlaceholderRenderingContext>(context))
-                continue;
-#endif
+            if (!is<CanvasRenderingContext2D>(context)
+#if ENABLE(OFFSCREEN_CANVAS)
+                && !is<OffscreenCanvasRenderingContext>(context)
+#endif
+                && !is<WebGLRenderingContextBase>(context)
+                && !is<ImageBitmapRenderingContext(context))
+                continue;
```
or
```
if (is<GPUCanvasContext>(context))
    continue;
```
because when the time comes to actually add back support for WebGPU in Web Inspector, it'll likely not be done using the `GPUCanvasContext` for the same reasons outlined in [215342@main](https://commits.webkit.org/215342@main)

Comment (mwyrzykowski) [reply Source/WebCore/inspector/InspectorCanvas.cpp:897]: Repro can be observed by removing these web inspector changes, opening https://webgpu.github.io/webgpu-samples/samples/worker, and then opening web inspector

Happy to consider other alternatives if there is a better option 👍

Comment (mwyrzykowski) [reply Source/WebCore/inspector/InspectorCanvas.cpp:897]: Oh great thank you @dcrousso! I’ll give that a try

Comment (dcrousso) [reply Source/WebCore/inspector/InspectorCanvas.cpp:897]: https://github.com/WebKit/WebKit/pull/19064
